dali/kernels/slice/slice_gpu.cuh

```diff
@@ -261,11 +261,15 @@ class SliceGPU {
       total_volume += volume(args.shape);
     }
 
-    unsigned min_blocks = 4 * GetSmCount();
-    block_size_ = kMaxBlockSize;
-    while (total_volume / block_size_ < min_blocks && block_size_ > kMinBlockSize) {
-      block_size_ /= 2;
-    }
+    int blocksPerSM;
+    CUDA_CALL(cudaOccupancyMaxActiveBlocksPerMultiprocessor(&blocksPerSM,
+              detail::SliceKernel<OutputType, InputType, Dims, false>, kBlockDim, 0));
+    unsigned blocks_in_wave = blocksPerSM * GetSmCount();
```

Comment:
What's a "wave"? This is a foreign term in our code.

Reply:
I've seen the term _wave_ used in Nsight Compute and some [NVIDIA's article](https://developer.nvidia.com/blog/cuda-pro-tip-minimize-the-tail-effect/), so I thought it is a common term. Sorry for that – there's a lot of terminology in CUDA and I still don't always know which terms are really common and which are some Nsight-specific words.

A _wave_ is a set of threads of specific kernel that can be run concurrently on a GPU (I think they call it so because when all threads take the same amount of time to complete, the first _wave_ is being run, then the second _wave_ is run after the first one finishes, so it looks like real-world waves, coming one after another)

So `blocks_in_wave` in the code above should probably be `max_active_blocks`, I'll change that in a moment.

And a side question:

In the PR description I use the terms (also taken from Nsight) _full wave_ to describe a _wave_ that fully utilizes the GPU, and a _partial wave_ to describe a _wave_ that leaves some part of the GPU idle, causing the tail effect. For example if the wave size for my kernel is 400 threads, and I run 401 threads where each takes 1s to complete, the first _full wave_ of 400 threads takes 1s to complete (which is good), and the second _partial wave_ of only one thread also takes 1s to complete, which is terrible.

Are there any common terms that I should use instead of _full wave_ and _partial wave_?

Reply:
> @hugo213 it seems it fails in:
> 
> ```
> test_operator_subscript.test_constant_ranges(<function test_constant_ranges.<locals>.data_gen at 0x7fbc182b1dc0>, 'AB', 'A', <function test_constant_ranges.<locals>.<lambda> at 0x7fbba9f83e50>, None) ... ok
> test_operator_subscript.test_swapped_ends ... test_body.sh: line 7: 30510 Floating point exception(core dumped) nosetests --verbose --attr '!slow,!pytorch,!mxnet,!cupy' ${test_script}
> ++ RV=136
> ```
> 
> and
> 
> ```
> test_operator_readers_numpy.test_numpy_reader_roi('empty range', '/scratch/4j1kxh1p', 8, 2, <class 'numpy.uint8'>, 'gpu', False, '*.npy') ... test_body.sh: line 11: 31422 Floating point exception(core dumped) nosetests --verbose --attr '!slow,!pytorch,!mxnet,!cupy' ${test_script}
> ++ RV=136
> ```
> 
> Could you check it?

Linux sometimes reports an integer division by zero as a "floating point exception". Can it be that you've hit a degenerate case and try to calculate parameters or something along these lines?